wallet/storage/src/internal/store_tx.rs

```diff
@@ -99,32 +142,89 @@ macro_rules! impl_read_ops {
             }
 
+            fn get_encryption_key_kdf_challenge(&self) -> crate::Result<Option<KdfChallenge>> {
+                self.read_value::<well_known::EncryptionKeyKdfChallenge>()
+            }
+
             fn get_root_key(&self, id: &RootKeyId) -> crate::Result<Option<RootKeyContent>> {
-                self.read::<db::DBRootKeys, _, _>(id)
+                utils::ensure!(!self.locked, crate::Error::WalletLocked());
+                Ok(self
+                    .read::<db::DBRootKeys, _, _>(id)?
+                    .map(|v| match self.encryption_key.as_ref() {
+                        Some(key) => {
+                            (key.decrypt(v.as_slice(), None)
+                                .expect("key has been checked in unlock_private_keys"))
+                        }
+                        None => (v),
+                    })
+                    .map(|v| RootKeyContent::decode_all(&mut v.as_slice()).expect("valid decode")))
             }
 
             /// Collect and return all keys from the storage
             fn get_all_root_keys(&self) -> crate::Result<BTreeMap<RootKeyId, RootKeyContent>> {
-                self.0
+                utils::ensure!(!self.locked, crate::Error::WalletLocked());
+                self.storage
                     .get::<db::DBRootKeys, _>()
                     .prefix_iter_decoded(&())
+                    .map_err(crate::Error::from)
+                    .map(|item| {
+                        item.map(|(k, v)| match self.encryption_key.as_ref() {
+                            Some(key) => (
+                                k,
+                                key.decrypt(v.as_slice(), None)
+                                    .expect("key has been checked in unlock_private_keys"),
+                            ),
+                            None => (k, v),
+                        })
+                        .map(|(k, v)| {
+                            (
+                                k,
+                                RootKeyContent::decode_all(&mut v.as_slice())
+                                    .expect("valid decode"),
+                            )
+                        })
+                    })
                     .map(Iterator::collect)
             }
 
+            /// Check if the provided encryption_key can decrypt all of the root keys
+            fn check_can_decrypt_all_root_keys(
+                &self,
+                encryption_key: &SymmetricKey,
+            ) -> crate::Result<()> {
+                self.storage
+                    .get::<db::DBRootKeys, _>()
+                    .prefix_iter_decoded(&())
+                    .map_err(crate::Error::from)
+                    .map(|mut item| {
+                        item.try_for_each(|(_, v)| {
+                            let decrypted_value = encryption_key
+                                .decrypt(v.as_slice(), None)
+                                .map_err(|_| crate::Error::WalletInvalidPassword())?;
```

Comment:
Not familiar with the encryption algorithm being used here. Does this have sufficient integrity checks to reliably detect the symmetric key used is the correct one?

Reply:
the algorithm is ChaCha20-Poly1305 which is an authenticated cipher, it will verify that the decryption is correct.